### ISSUE TITLE
boolean value change in function is not considered by TS in control flow

### ISSUE THREAD

**Issue by paradite** [microsoft/TypeScript#37486]
**TypeScript Version:**  3.8.3

**Search Terms:** This condition will always return 

**Expected behavior:** compiler does not report error

**Actual behavior:** compiler reports error `This condition will always return 'false' since the types 'false' and 'true' have no overlap.(2367)`

**Related Issues:** NA

**Code**
```ts
let a: boolean = false;
let b: boolean = false;

const run = () => {
  a = true;
};

function run2() {
  b = true;
}

run();
run2();

if (a === true || b === true) {
  console.log(a);
  console.log(b);
}

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
let a = false;
let b = false;
const run = () => {
    a = true;
};
function run2() {
    b = true;
}
run();
run2();
if (a === true || b === true) {
    console.log(a);
    console.log(b);
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=2&pc=22#code/DYUwLgBAhgXBBGB7RooDsIF4IDMrAGcQBuAKFEnjiRRHS13yLNIGNE0DIAnAVw2wAKAJRYAfBADepCNAZg+JUgF8WOfqzABLDhD5oATCKkyE8xWWWlS+kWX1HhLLTgiCoWTNgW8QEAD7+Zl7eiqLSsuyctAB0wIgA5u5OplEEsfFJ8ClWQA)

**Comment by IllusionMH** [comment]
This is design limitation. See #9998 for more details.
